---
title: 'Sheet 3: Phyloseq'
author: "Ezequiel Ibarra"
date: '`r Sys.Date()`'
output: html_document
---

## IMPORT PACKAGES

```{r, include = FALSE} 
library(phyloseq) 
library(plyr)
library(ggpubr)
library(kableExtra)
library(microbiome)
```

## READ OBJECTS

```{r, include = FALSE} 
seqtab.nochim <- readRDS("objects/Sheet1/seqtab.nochim.rds")
samdf <- readRDS("objects/Sheet2/samdf.rds")
taxa <- readRDS("objects/Sheet2/taxa.rds")
fitGTR <- readRDS("objects/Sheet2/fitGTR.rds")
```

## FIRST OBJECT PHYLOSEQ

The phyloseq R package is a powerful framework for further analysis of microbiome data. We now demonstrate how to straightforwardly import the tables produced by the DADA2 pipeline into phyloseq.

Construct a phyloseq object directly from the dada2 outputs.

```{r, include = FALSE}
ps1 <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
                sample_data(samdf), 
                tax_table(taxa),
                phy_tree(fitGTR$tree))
dna <- Biostrings::DNAStringSet(taxa_names(ps1))
names(dna) <- taxa_names(ps1)
ps1 <- merge_phyloseq(ps1, dna)
taxa_names(ps1) <- paste0("ASV", seq(ntaxa(ps1)))
ps1 
```


## QUALITY CONTROL

```{r, include = FALSE}
prevdf = apply(X = otu_table(ps1),
               MARGIN = ifelse(taxa_are_rows(ps1), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})
#prevdf
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps1),
                    tax_table(ps1))

plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))}) -> dfprev
kable(dfprev)
```

## NON-BACTERIAL SEQUENCE FILTER

```{r, include = FALSE}
filterPhyla2 <- c("Chloroplast", "Mitochondria", "Eukaryota")
psd1 <- subset_taxa(ps1, !Kingdom %in% filterPhyla2)
psd1 <- subset_taxa(ps1, !Phylum %in% filterPhyla2)
psd1 <- subset_taxa(ps1, !Class %in% filterPhyla2)
psd1 <- subset_taxa(ps1, !Order %in% filterPhyla2)
psd1 <- subset_taxa(ps1, !Family %in% filterPhyla2)
psd1 <- subset_taxa(ps1, !Genus %in% filterPhyla2)
```

## ADDITIONAL FILTERS (NUMBER OF READINGS AND PERCENTAGE PER SAMPLE)

```{r, warning = FALSE, include = FALSE}
psd2 <- filter_taxa(psd1, function(x) mean(x) > 1e-5, TRUE)
psd3 <- filter_taxa(psd2, function(x) sum(x > 2) > (0.1*length(x)), TRUE)
psd4 = prune_samples(sample_sums(psd3) > 1000, psd3)
```

## INTEREST RATE

```{r, warning=FALSE, fig.align='center'}
prevdf1 = subset(prevdf, Phylum %in% get_taxa_unique(psd4, "Phylum"))
ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(ps1),color=Phylum)) +
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position="none", legend.text = element_text(face = "italic"))
(prevalenceThreshold = 0.05 * nsamples(psd4))
keepTaxa = rownames(prevdf1)[(prevdf1$Prevalence >= prevalenceThreshold)]
(psd5 = prune_taxa(keepTaxa, psd4))
taxa_names(psd5) <- paste0("ASV", seq(ntaxa(psd5)))
sample_sum_df <- data.frame(sum = sample_sums(psd5))

# GRAPH (Distribution of sample sequencing depth)
ggplot(sample_sum_df, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "grey", binwidth = 2500) +
  ggtitle("Distribution of sample sequencing depth") + 
  xlab("Read counts") +
  theme(axis.title.y = element_blank())

scripts <- c("graphical_methods.R",
             "tree_methods.R",
             "plot_merged_trees.R",
             "specificity_methods.R",
             "ternary_plot.R",
             "richness.R",
             "edgePCA.R",
             "copy_number_correction.R",
             "import_frogs.R",
             "prevalence.R",
             "compute_niche.R")
urls <- paste0("https://raw.githubusercontent.com/mahendra-mariadassou/phyloseq-extended/master/R/", scripts)

for (url in urls) {
  source(url)
}
```

## RAREFACTION CURVES

```{r, warning=FALSE, fig.align='center'}
p <- ggrare(psd5, step = 100, color = "Treatment", label = rownames(psd5@sam_data), se = TRUE)
(p <- p + facet_wrap(~Treatment))
```

## PHYLOGENY TREE GRAPH

```{r}
plot_tree(psd5, method = "treeonly", ladderize = "left")
summarize_phyloseq(psd5)
df <- psmelt(psd5)
```

## PERCENTAGE

```{r}
psd5r  = transform_sample_counts(psd5, function(x) x / sum(x) )
```

